File: qt/multimedia/gen_qscreencapture.cpp
// +build ignore

#include <QMediaCaptureSession>
#include <QMetaObject>
#include <QObject>
#include <QScreen>
#include <QScreenCapture>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qscreencapture.h>
#include "gen_qscreencapture.h"

class MiqtVirtualQScreenCapture : public virtual QScreenCapture {
public:

	MiqtVirtualQScreenCapture(): QScreenCapture() {};
	MiqtVirtualQScreenCapture(QObject* parent): QScreenCapture(parent) {};

	virtual ~MiqtVirtualQScreenCapture() = default;

	// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;

	// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
		if (handle__MetaObject == 0) {
			return QScreenCapture::metaObject();
		}
		

		QMetaObject* callback_return_value = miqt_exec_callback_QScreenCapture_MetaObject(const_cast<MiqtVirtualQScreenCapture*>(this), handle__MetaObject);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QScreenCapture::metaObject();

	}

	// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;

	// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
		if (handle__Metacast == 0) {
			return QScreenCapture::qt_metacast(param1);
		}
		
		const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QScreenCapture_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QScreenCapture::qt_metacast(param1);

	}

};

QScreenCapture* QScreenCapture_new() {
	return new MiqtVirtualQScreenCapture();
}

QScreenCapture* QScreenCapture_new2(QObject* parent) {
	return new MiqtVirtualQScreenCapture(parent);
}

void QScreenCapture_virtbase(QScreenCapture* src, QObject** outptr_QObject) {
	*outptr_QObject = static_cast<QObject*>(src);
}

QMetaObject* QScreenCapture_MetaObject(const QScreenCapture* self) {
	return (QMetaObject*) self->metaObject();
}

void* QScreenCapture_Metacast(QScreenCapture* self, const char* param1) {
	return self->qt_metacast(param1);
}

struct miqt_string QScreenCapture_Tr(const char* s) {
	QString _ret = QScreenCapture::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

QMediaCaptureSession* QScreenCapture_CaptureSession(const QScreenCapture* self) {
	return self->captureSession();
}

void QScreenCapture_SetScreen(QScreenCapture* self, QScreen* screen) {
	self->setScreen(screen);
}

QScreen* QScreenCapture_Screen(const QScreenCapture* self) {
	return self->screen();
}

bool QScreenCapture_IsActive(const QScreenCapture* self) {
	return self->isActive();
}

Error QScreenCapture_Error(const QScreenCapture* self) {
	return self->error();
}

struct miqt_string QScreenCapture_ErrorString(const QScreenCapture* self) {
	QString _ret = self->errorString();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QScreenCapture_SetActive(QScreenCapture* self, bool active) {
	self->setActive(active);
}

void QScreenCapture_Start(QScreenCapture* self) {
	self->start();
}

void QScreenCapture_Stop(QScreenCapture* self) {
	self->stop();
}

void QScreenCapture_ActiveChanged(QScreenCapture* self, bool param1) {
	self->activeChanged(param1);
}

void QScreenCapture_connect_ActiveChanged(QScreenCapture* self, intptr_t slot) {
	MiqtVirtualQScreenCapture::connect(self, static_cast<void (QScreenCapture::*)(bool)>(&QScreenCapture::activeChanged), self, [=](bool param1) {
		bool sigval1 = param1;
		miqt_exec_callback_QScreenCapture_ActiveChanged(slot, sigval1);
	});
}

void QScreenCapture_ErrorChanged(QScreenCapture* self) {
	self->errorChanged();
}

void QScreenCapture_connect_ErrorChanged(QScreenCapture* self, intptr_t slot) {
	MiqtVirtualQScreenCapture::connect(self, static_cast<void (QScreenCapture::*)()>(&QScreenCapture::errorChanged), self, [=]() {
		miqt_exec_callback_QScreenCapture_ErrorChanged(slot);
	});
}

void QScreenCapture_ScreenChanged(QScreenCapture* self, QScreen* param1) {
	self->screenChanged(param1);
}

void QScreenCapture_connect_ScreenChanged(QScreenCapture* self, intptr_t slot) {
	MiqtVirtualQScreenCapture::connect(self, static_cast<void (QScreenCapture::*)(QScreen*)>(&QScreenCapture::screenChanged), self, [=](QScreen* param1) {
		QScreen* sigval1 = param1;
		miqt_exec_callback_QScreenCapture_ScreenChanged(slot, sigval1);
	});
}

void QScreenCapture_ErrorOccurred(QScreenCapture* self, int error, struct miqt_string errorString) {
	QString errorString_QString = QString::fromUtf8(errorString.data, errorString.len);
	self->errorOccurred(static_cast<QScreenCapture::Error>(error), errorString_QString);
}

void QScreenCapture_connect_ErrorOccurred(QScreenCapture* self, intptr_t slot) {
	MiqtVirtualQScreenCapture::connect(self, static_cast<void (QScreenCapture::*)(QScreenCapture::Error, const QString&)>(&QScreenCapture::errorOccurred), self, [=](QScreenCapture::Error error, const QString& errorString) {
		QScreenCapture::Error error_ret = error;
		int sigval1 = static_cast<int>(error_ret);
		const QString errorString_ret = errorString;
		// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
		QByteArray errorString_b = errorString_ret.toUtf8();
		struct miqt_string errorString_ms;
		errorString_ms.len = errorString_b.length();
		errorString_ms.data = static_cast<char*>(malloc(errorString_ms.len));
		memcpy(errorString_ms.data, errorString_b.data(), errorString_ms.len);
		struct miqt_string sigval2 = errorString_ms;
		miqt_exec_callback_QScreenCapture_ErrorOccurred(slot, sigval1, sigval2);
	});
}

struct miqt_string QScreenCapture_Tr2(const char* s, const char* c) {
	QString _ret = QScreenCapture::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QScreenCapture_Tr3(const char* s, const char* c, int n) {
	QString _ret = QScreenCapture::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QScreenCapture_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQScreenCapture*>( (QScreenCapture*)(self) )->handle__MetaObject = slot;
}

QMetaObject* QScreenCapture_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQScreenCapture*)(self) )->virtualbase_MetaObject();
}

void QScreenCapture_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQScreenCapture*>( (QScreenCapture*)(self) )->handle__Metacast = slot;
}

void* QScreenCapture_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQScreenCapture*)(self) )->virtualbase_Metacast(param1);
}

void QScreenCapture_Delete(QScreenCapture* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQScreenCapture*>( self );
	} else {
		delete self;
	}
}

File: qt/multimedia/gen_qscreencapture.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QScreenCapture__Error int

const (
	QScreenCapture__NoError               QScreenCapture__Error = 0
	QScreenCapture__InternalError         QScreenCapture__Error = 1
	QScreenCapture__CapturingNotSupported QScreenCapture__Error = 2
	QScreenCapture__CaptureFailed         QScreenCapture__Error = 4
	QScreenCapture__NotFound              QScreenCapture__Error = 5
)

type QScreenCapture struct {
	h          uintptr
	isSubclass bool
}

// NewQScreenCapture constructs a new QScreenCapture object.
func NewQScreenCapture() *QScreenCapture {
	g := newQScreenCapture(QScreenCapture_new())
	g.isSubclass = true
	return g
}

// NewQScreenCapture2 constructs a new QScreenCapture object.
func NewQScreenCapture2(parent *qt.QObject) *QScreenCapture {
	g := newQScreenCapture(QScreenCapture_new2((*QObject)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

func (this *QScreenCapture) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QScreenCapture_MetaObject(this.h)))
}

func (this *QScreenCapture) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QScreenCapture_Metacast(this.h, param1_Cstring))
}

func QScreenCapture_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QScreenCapture_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QScreenCapture) CaptureSession() *QMediaCaptureSession {
	return newQMediaCaptureSession(QScreenCapture_CaptureSession(this.h))
}

func (this *QScreenCapture) SetScreen(screen *qt.QScreen) {
	QScreenCapture_SetScreen(this.h, (*QScreen)(screen.UnsafePointer()))
}

func (this *QScreenCapture) Screen() *qt.QScreen {
	return qt.UnsafeNewQScreen(unsafe.Pointer(QScreenCapture_Screen(this.h)))
}

func (this *QScreenCapture) IsActive() bool {
	return (bool)(QScreenCapture_IsActive(this.h))
}

func (this *QScreenCapture) Error() Error {
	xxxxxxxxx
}

func (this *QScreenCapture) ErrorString() string {
	var _ms struct_miqt_string = QScreenCapture_ErrorString(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QScreenCapture) SetActive(active bool) {
	QScreenCapture_SetActive(this.h, (bool)(active))
}

func (this *QScreenCapture) Start() {
	QScreenCapture_Start(this.h)
}

func (this *QScreenCapture) Stop() {
	QScreenCapture_Stop(this.h)
}

func (this *QScreenCapture) ActiveChanged(param1 bool) {
	QScreenCapture_ActiveChanged(this.h, (bool)(param1))
}

func (this *QScreenCapture) OnActiveChanged(slot func(param1 bool)) {
	QScreenCapture_connect_ActiveChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QScreenCapture_ActiveChanged
func miqt_exec_callback_QScreenCapture_ActiveChanged(cb intptr_t, param1 bool) {
	gofunc, ok := cgo.Handle(cb).Value().(func(param1 bool))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (bool)(param1)

	gofunc(slotval1)
}

func (this *QScreenCapture) ErrorChanged() {
	QScreenCapture_ErrorChanged(this.h)
}

func (this *QScreenCapture) OnErrorChanged(slot func()) {
	QScreenCapture_connect_ErrorChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QScreenCapture_ErrorChanged
func miqt_exec_callback_QScreenCapture_ErrorChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QScreenCapture) ScreenChanged(param1 *qt.QScreen) {
	QScreenCapture_ScreenChanged(this.h, (*QScreen)(param1.UnsafePointer()))
}

func (this *QScreenCapture) OnScreenChanged(slot func(param1 *qt.QScreen)) {
	QScreenCapture_connect_ScreenChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QScreenCapture_ScreenChanged
func miqt_exec_callback_QScreenCapture_ScreenChanged(cb intptr_t, param1 *QScreen) {
	gofunc, ok := cgo.Handle(cb).Value().(func(param1 *qt.QScreen))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := qt.UnsafeNewQScreen(unsafe.Pointer(param1))

	gofunc(slotval1)
}

func (this *QScreenCapture) ErrorOccurred(error QScreenCapture__Error, errorString string) {
	errorString_ms := struct_miqt_string{}
	errorString_ms.data = CString(errorString)
	errorString_ms.len = size_t(len(errorString))
	defer free(unsafe.Pointer(errorString_ms.data))
	QScreenCapture_ErrorOccurred(this.h, (int)(error), errorString_ms)
}

func (this *QScreenCapture) OnErrorOccurred(slot func(error QScreenCapture__Error, errorString string)) {
	QScreenCapture_connect_ErrorOccurred(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QScreenCapture_ErrorOccurred
func miqt_exec_callback_QScreenCapture_ErrorOccurred(cb intptr_t, error int, errorString struct_miqt_string) {
	gofunc, ok := cgo.Handle(cb).Value().(func(error QScreenCapture__Error, errorString string))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (QScreenCapture__Error)(error)

	var errorString_ms struct_miqt_string = errorString
	errorString_ret := GoStringN(errorString_ms.data, int(int64(errorString_ms.len)))
	free(unsafe.Pointer(errorString_ms.data))
	slotval2 := errorString_ret

	gofunc(slotval1, slotval2)
}

func QScreenCapture_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QScreenCapture_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QScreenCapture_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QScreenCapture_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QScreenCapture) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QScreenCapture_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QScreenCapture) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QScreenCapture_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QScreenCapture_MetaObject
func miqt_exec_callback_QScreenCapture_MetaObject(self QScreenCapture, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QScreenCapture{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QScreenCapture) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QScreenCapture_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QScreenCapture) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QScreenCapture_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QScreenCapture_Metacast
func miqt_exec_callback_QScreenCapture_Metacast(self QScreenCapture, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QScreenCapture{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
